File: LeetCode/Problem18_4Sum/4Sum.go
func fourSum(nums []int, target int) [][]int {
	n := len(nums)
	sort.Ints(nums)
	pairMap := map[int][][]int{}
	result := map[[4]int]struct{}{}

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			sum := nums[i] + nums[j]
			pairMap[sum] = append(pairMap[sum], []int{i, j})
		}
	}

	for sum1, pairs1 := range pairMap {
		sum2 := target - sum1
		pairs2, ok := pairMap[sum2]
		if !ok {
			continue
		}
		for _, p1 := range pairs1 {
			for _, p2 := range pairs2 {
				i, j, k, l := p1[0], p1[1], p2[0], p2[1]
				if i == k || i == l || j == k || j == l {
					continue
				}
				quad := []int{nums[i], nums[j], nums[k], nums[l]}
				sort.Ints(quad)
				result[[4]int{quad[0], quad[1], quad[2], quad[3]}] = struct{}{}
			}
		}
	}

	res := [][]int{}
	for k := range result {
		res = append(res, []int{k[0], k[1], k[2], k[3]})
	}
	return res
}

File: LeetCode/Problem18_4Sum/4Sum.cpp
class Solution {
  public:
      vector<vector<int>> fourSum(vector<int>& nums, int target) {
          int n = nums.size();
          unordered_map<int, vector<pair<int, int>>> twoSum;
          set<vector<int>> result;
          sort(nums.begin(), nums.end());
  
          for (int i = 0; i < n; ++i)
              for (int j = i + 1; j < n; ++j)
                  twoSum[nums[i] + nums[j]].emplace_back(i, j);
  
          for (auto& [sum1, pairs1] : twoSum) {
              int sum2 = target - sum1;
              if (!twoSum.count(sum2)) continue;
  
              for (auto [i, j] : pairs1) {
                  for (auto [k, l] : twoSum[sum2]) {
                      set<int> s = {i, j, k, l};
                      if (s.size() < 4) continue;
  
                      vector<int> quad = {nums[i], nums[j], nums[k], nums[l]};
                      sort(quad.begin(), quad.end());
                      result.insert(quad);
                  }
              }
          }
  
          return vector<vector<int>>(result.begin(), result.end());
      }
  };
  